pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,346 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/pkg/testutil"
+)
+
+const (
+	testAppID     = "asdf"
+	testVersion   = "1.0.0"
+	testInstallID = "yv66vt6tvu8="
+	testServerURL = "https://example.com"
+)
+
+func defaultClient() *client {
+	return &client{
+		AppID:      testAppID,
+		Version:    testVersion,
+		InstallID:  testInstallID,
+		HTTPClient: &http.Client{Timeout: 1 * time.Second},
+		OptOut:     false,
+		Config: &metricsConfig{
+			ServerURL: testServerURL,
+			NoMetrics: false,
+		},
+	}
+}
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+	t.Run("happy_path", func(t *testing.T) {
+		t.Parallel()
+
+		cases := []struct {
+			name      string
+			client    *http.Client
+			installID string
+			want      *client
+		}{
+			{
+				name: "happy_path_no_install_id",
+				want: defaultClient(),
+			},
+			{
+				name:      "happy_path_with_install_id",
+				installID: testInstallID,
+				want:      defaultClient(),
+			},
+			{
+				name:      "happy_path_with_custom_http_client",
+				installID: testInstallID,
+				client:    &http.Client{Timeout: 2},
+				want: func() *client {
+					c := defaultClient()
+					c.HTTPClient = &http.Client{Timeout: 2}
+					return c
+				}(),
+			},
+		}
+
+		for _, tc := range cases {
+			t.Run(tc.name, func(t *testing.T) {
+				t.Parallel()
+
+				ctx := context.Background()
+
+				installPath := t.TempDir() + "/" + installIDFileName
+				if tc.installID != "" {
+					if err := storeInstallID(testAppID, installPath, &InstallIDData{tc.installID}); err != nil {
+						t.Fatalf("test setup failed: %s", err.Error())
+					}
+				}
+				envVars := map[string]string{
+					"METRICS_URL": testServerURL,
+				}
+				lookuper := envconfig.MapLookuper(envVars)
+				opts := []Option{
+					WithLookuper(lookuper),
+					WithInstallIDFileOverride(installPath),
+				}
+				if tc.client != nil {
+					opts = append(opts, WithHTTPClient(tc.client))
+				}
+
+				i, err := New(ctx, testAppID, testVersion, opts...)
+				if err != nil {
+					t.Errorf("unexpected error: %s", err.Error())
+				}
+				got := i.(*client) //nolint:forcetypeassert
```

Comment:
It'd probably be better to do the type check

```suggestion
				got, ok := i.(*client) //nolint:forcetypeassert
				if !ok {
				    t.Fatal(...)
				}
```

Reply:
I see, I guess I figured I would just let it panic, but I'll change those.

Reply:
Left one in testcase definition as it would no longer be an easy one-liner (would need a closure or something)